litmus-portal/cluster-agents/subscriber/pkg/workloads/workloads.go

```diff
@@ -0,0 +1,59 @@
+package workloads
+
+import (
+	"context"
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func GetPodsFromWorkload(ns, kind string, allPods *corev1.PodList, wld []string, client *kubernetes.Clientset, getPodOwner func(ns, kind string, wld []string, client *kubernetes.Clientset) ([]string, string, error)) ([]string, error) {
+
+	ownerName, ownerKind, err := getPodOwner(ns, kind, wld, client)
+	if err != nil {
+		return nil, err
+	}
+
+	var pods []string
+	for _, r := range allPods.Items {
+		if checkOwnerRef(r.OwnerReferences, ownerName, ownerKind) {
+			pods = append(pods, r.Name)
+		}
+	}
+	return pods, nil
+}
+
+func GetPodsFromServices(ns string, wld []string, client *kubernetes.Clientset) ([]string, error) {
+	var pods []string
+	for _, svcName := range wld {
+		svc, err := client.CoreV1().Services(ns).Get(context.Background(), svcName, v1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range svc.Spec.Selector {
+			res, err := client.CoreV1().Pods(svc.Namespace).List(context.Background(), v1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", k, v)})
+			if err != nil {
+				return nil, err
+			}
+			for _, p := range res.Items {
+				pods = append(pods, p.Name)
+			}
+		}
+	}
+	return pods, nil
+}
+
+var managedByReplicaSet = func(ns, kind string, wld []string, client *kubernetes.Clientset) ([]string, string, error) {
+	rs, err := getReplicaSetsFromWorkload(ns, kind, wld, client)
+	return rs, ReplicaSet, err
+}
+
+var managedByReplicaController = func(ns, kind string, wld []string, client *kubernetes.Clientset) ([]string, string, error) {
+	rc, err := getReplicaControllersFromWorkload(ns, kind, wld, client)
+	return rc, ReplicationController, err
+}
```

Comment:
This code including the logic to fetch and match owners for each of replicaset and replica controller can be done in a single function if we use dynamic client, providing the proper gvr fopr each type we would get an `unstructured.Unstructured` response from there we can get the ownerref using `GetOwnerReferences` which is all we care about as far as I can see. That would make the code much more clean and reusable.

Reply:
gvr  & dc can be used better if we traverse from parent to child. But we are deriving OwnerRef from child to parent `pod` --> `rs/rc/ds/sts` --> `dc/deploy/rollout`

But I will see if I can use it for sub-tasks

Reply:
I was thinking of something like this :
```
func getChildResourceFromParentresource(ns, kind string, wld []string, dynamicClient dynamic.Interface, childGVR schema.GroupVersionResource) ([]string, error) {

	resources, err := dynamicClient.Resource(childGVR).Namespace(ns).List(context.Background(), metav1.ListOptions{})
	if err != nil {
		return nil, err
	}

	var rs []string
	for _, r := range resources.Items {
		if checkOwnerRef(r.GetOwnerReferences(), wld, kind) {
			rs = append(rs, r.GetName())
		}
	}
	return rs, nil
}
```

Then you can just have one single function instead of having separate one for getting replicaset/replicacontroller/pods from their parent resources